Page views: use origin URL instead of page name

stsewd opened the pull request:
Page name is a sphinx only concept,
so we don't have correct data for page views from mkdocs projects.

Using the origin url + the unresolver gives us the correct path.
Also, we are always storing the full path (index.html instead of "/"),
so we don't have duplicates (let me know if we do want to have duplicates).

closes https://github.com/readthedocs/readthedocs.org/issues/7131

stsewd commented on readthedocs/api/v2/views/footer_views.py:
```diff
@@ -81,6 +81,19 @@ class BaseFooterHTML(APIView):
+    - docroot: Path where all the source documents are.
+      Used to build the ``edit_on`` URL.
+    - source_suffix: Suffix from the source document.
+      Used to build the ``edit_on`` URL.
```
We are passing subproject too, but we are not using it. I can remove it in another PR if that's ok

stsewd commented on readthedocs/rtd_tests/tests/test_footer.py:
```diff
@@ -418,7 +419,7 @@ class TestFooterPerformance(APITestCase):
     # The expected number of queries for generating the footer
     # This shouldn't increase unless we modify the footer API
-    EXPECTED_QUERIES = 14
+    EXPECTED_QUERIES = 13
```
We have one query less bc we are not sending the `origin` parameter, so it defaults to false before doing a query for the feature flag.

ericholscher replied:
We should test this with an origin. I expect the `unresolve` method to actually add a few queries, which we should be careful about :/

ericholscher commented on readthedocs/analytics/signals.py:
```diff
@@ -17,16 +18,22 @@ def increase_page_view_count(sender, **kwargs):
+    if not origin or not project.has_feature(Feature.STORE_PAGEVIEWS):
+        return
+
+    unresolved = unresolve(origin)
```
I'm not sure if this is the best approach right now. I think this is going to add a good number of queries to the footer, which is already our most expensive view. If we can just pull the path off the request, I think that is probably best, but maybe calling `unresolve` is correct, and we should focus on optimizations there?

stsewd commented:
Ok, I added a new method `unresolve_from_request`, since when we are in proxito (proxied footer) the request is passed through our middlewares that already set the request.slug attribute and everything else that the unresolver needs.

I updated the tests to run the whole request instead of just the footer view, that's why we have more queries. But the unresolver is only adding  one query. We may be seeing more queries for translations/subprojects, but that's in general, not only for the unresolver, I'll add more tests later.

humitos commented on readthedocs/api/v2/views/footer_views.py:
```diff
@@ -81,6 +81,19 @@ class BaseFooterHTML(APIView):
+    - version
+    - page: Sphinx's page name, used for path operations,
+      like change between languages (deprecated in favor of ``origin``).
+    - origin: Full path with domain, used for path operations.
```
Why don't we call it `full_path`?

humitos replied:
or `absolute_uri` as Django calls it when it has the domain on it.

stsewd replied:
I'm avoiding naming this path since it includes the domain as well, we use full path to refer to a path in other parts of the code. Not sure about uri

ericholscher replied:
Yea, origin isn't the right name, since that's used in HTTP for just the origin hostname: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin

I think `absolute_uri` or similar is best: https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.HttpRequest.build_absolute_uri

ericholscher commented on readthedocs/core/unresolver.py:
```diff
@@ -27,17 +27,35 @@ def unresolve(self, url):
+        """
+        parsed = urlparse(path)
+        path = parsed.path
+        project_slug = getattr(request, 'slug', None)
```
Believe we're using `host_project_slug` for this, `slug` is deprecated.

stsewd commented:
>  It will change how the data is stored for users, but thats probably fine?

Yeah, we will have duplicates from previous pages, like `index.html` and `/`. But it would settle down in a month or so after previous page views are deleted.
